event/bus.go

```diff
@@ -0,0 +1,179 @@
+// Copyright 2024 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package event
+
+import (
+	"sync/atomic"
+)
+
+type subscribeCommand *Subscription
+
+type unsubscribeCommand *Subscription
+
+type publishCommand Message
+
+type closeCommand struct{}
+
+// Message contains event info.
+type Message struct {
+	// Name is the name of the event this message was generated from.
+	Name string
+	// Data contains optional event information.
+	Data any
+}
+
+// NewMessage returns a new message with the given name and optional data.
+func NewMessage(name string, data any) Message {
+	return Message{name, data}
+}
+
+// Subscription is a read-only event stream.
+type Subscription struct {
+	id     uint64
+	value  chan Message
+	events []string
+}
+
+// Message returns the next event value from the subscription.
+func (s *Subscription) Message() <-chan Message {
+	return s.value
+}
+
+// Events returns the names of all subscribed events.
+func (s *Subscription) Events() []string {
+	return s.events
+}
+
+// Bus is used to broadcast events to subscribers.
+type Bus struct {
```

Comment:
todo: Doesn't have to be in this PR, but the `OnFooAsync` funcs added in https://github.com/sourcenetwork/defradb/pull/2708 can now be removed IMO - they are dead code (unless we want to keep them to allow users to call them).

Reply:
I'll try to remember to bring this up in the stand up